<a href="https://colab.research.google.com/github/DAIvdetango/Data-Analysis--Online-School-CRM/blob/main/Cleaning_data_done_ipynb%22%22.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

[![Open in Colab](http://colab.reseach.google.com/assets/colab-badge.svg)]

In [ ]:
import matplotlib.pyplot as plt
import pandas as pd
import numpy as np
import seaborn as sns
import datetime

In [ ]:
def load_data(file_path: str, file_type: str = "csv", **kwargs) -> pd.DataFrame:
    if file_type == "csv":
        df = pd.read_csv(file_path, **kwargs)
    elif file_type in ("excel", "xlsx"):
        df = pd.read_excel(file_path, **kwargs)
    else:
        raise ValueError(f"Unsupported file type: {file_type}")
    return df


##✨ **Data Cleaning `contacts_df`**

In [ ]:
df_path = '/content/Contacts (Done).xlsx'
contacts_df = load_data(df_path, 'xlsx', dtype={'Id': str})
contacts_df.head()

Id Contact Owner Name      Created Time     Modified Time
0  5805028000000645014       Rachel White  27.06.2023 11:28  22.12.2023 13:34
1  5805028000000872003      Charlie Davis  03.07.2023 11:31  21.05.2024 10:23
2  5805028000000889001          Bob Brown  02.07.2023 22:37  21.12.2023 13:17
3  5805028000000907006          Bob Brown  03.07.2023 05:44  29.12.2023 15:20
4  5805028000000939010         Nina Scott  04.07.2023 10:11  16.04.2024 16:14

In [ ]:
contacts_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 18548 entries, 0 to 18547
Data columns (total 4 columns):
 #   Column              Non-Null Count  Dtype 
---  ------              --------------  ----- 
 0   Id                  18548 non-null  object
 1   Contact Owner Name  18548 non-null  object
 2   Created Time        18548 non-null  object
 3   Modified Time       18548 non-null  object
dtypes: object(4)
memory usage: 579.8+ KB


In [ ]:
contacts_df.describe()

Id Contact Owner Name      Created Time  \
count                 18548              18548             18548   
unique                18548                 28             17921   
top     5805028000000645014      Charlie Davis  10.06.2024 09:00   
freq                      1               2018                13   

           Modified Time  
count              18548  
unique             16580  
top     13.06.2024 17:08  
freq                  25

---
##  Data Type Conversions
---

In [ ]:
contacts_df["Created Time"] = pd.to_datetime(contacts_df["Created Time"], format='%d.%m.%Y %H:%M', dayfirst=True, errors="raise")
contacts_df["Modified Time"] = pd.to_datetime(contacts_df["Modified Time"], format='%d.%m.%Y %H:%M', dayfirst=True, errors="raise")

In [ ]:
contacts_df.describe()

Created Time                  Modified Time
count                          18548                          18548
mean   2024-01-24 14:00:21.679965696  2024-02-15 07:41:24.814535168
min              2023-06-27 11:28:00            2023-07-06 10:54:00
25%              2023-11-15 16:40:45            2023-12-09 12:43:00
50%              2024-02-01 18:30:00            2024-02-28 22:21:30
75%              2024-04-12 16:37:15            2024-04-26 22:02:30
max              2024-06-21 15:30:00            2024-06-21 15:32:00

In [ ]:
contacts_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 18548 entries, 0 to 18547
Data columns (total 4 columns):
 #   Column              Non-Null Count  Dtype         
---  ------              --------------  -----         
 0   Id                  18548 non-null  object        
 1   Contact Owner Name  18548 non-null  object        
 2   Created Time        18548 non-null  datetime64[ns]
 3   Modified Time       18548 non-null  datetime64[ns]
dtypes: datetime64[ns](2), object(2)
memory usage: 579.8+ KB


---
## Removing duplicates
---

In [ ]:
contacts_df.duplicated(subset=contacts_df.columns[1:]).sum()

38

In [ ]:
contacts_df[contacts_df.duplicated(keep=False, subset=contacts_df.columns[1:])]

Id Contact Owner Name        Created Time  \
203    5805028000001949074          Bob Brown 2023-07-17 18:40:00   
205    5805028000001953081          Bob Brown 2023-07-17 18:40:00   
279    5805028000002340967          Bob Brown 2023-07-18 16:53:00   
280    5805028000002344049          Bob Brown 2023-07-18 16:53:00   
334    5805028000002740077          Bob Brown 2023-07-21 12:26:00   
...                    ...                ...                 ...   
17864  5805028000054231884       Rachel White 2024-06-10 09:00:00   
17865  5805028000054232018       Rachel White 2024-06-10 09:00:00   
17869  5805028000054238271       Rachel White 2024-06-10 09:00:00   
17870  5805028000054238317       Rachel White 2024-06-10 09:00:00   
17872  5805028000054238365       Rachel White 2024-06-10 09:00:00   

            Modified Time  
203   2023-07-17 18:40:00  
205   2023-07-17 18:40:00  
279   2023-07-18 16:53:00  
280   2023-07-18 16:53:00  
334   2023-07-21 12:26:00  
...                   ...  
17864 2024-06-10 09:33:00  
17865 2024-06-10 09:33:00  
17869 2024-06-10 09:33:00  
17870 2024-06-10 09:33:00  
17872 2024-06-10 09:33:00  

[71 rows x 4 columns]

In [ ]:
contacts_df.drop_duplicates(subset=contacts_df.columns[1:], inplace=True)

contacts_df[contacts_df.duplicated(subset=contacts_df.columns[1:])].shape

(0, 4)

---
## Processing Missing Values
---

In [ ]:
contacts_df.isna().sum()

Id                    0
Contact Owner Name    0
Created Time          0
Modified Time         0
dtype: int64

---
## Results Output
---

In [ ]:
contacts_df.head()

Id Contact Owner Name        Created Time  \
0  5805028000000645014       Rachel White 2023-06-27 11:28:00   
1  5805028000000872003      Charlie Davis 2023-07-03 11:31:00   
2  5805028000000889001          Bob Brown 2023-07-02 22:37:00   
3  5805028000000907006          Bob Brown 2023-07-03 05:44:00   
4  5805028000000939010         Nina Scott 2023-07-04 10:11:00   

        Modified Time  
0 2023-12-22 13:34:00  
1 2024-05-21 10:23:00  
2 2023-12-21 13:17:00  
3 2023-12-29 15:20:00  
4 2024-04-16 16:14:00

In [ ]:
contacts_df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 18510 entries, 0 to 18547
Data columns (total 4 columns):
 #   Column              Non-Null Count  Dtype         
---  ------              --------------  -----         
 0   Id                  18510 non-null  object        
 1   Contact Owner Name  18510 non-null  object        
 2   Created Time        18510 non-null  datetime64[ns]
 3   Modified Time       18510 non-null  datetime64[ns]
dtypes: datetime64[ns](2), object(2)
memory usage: 723.0+ KB


In [ ]:
contacts_df

Id Contact Owner Name        Created Time  \
0      5805028000000645014       Rachel White 2023-06-27 11:28:00   
1      5805028000000872003      Charlie Davis 2023-07-03 11:31:00   
2      5805028000000889001          Bob Brown 2023-07-02 22:37:00   
3      5805028000000907006          Bob Brown 2023-07-03 05:44:00   
4      5805028000000939010         Nina Scott 2023-07-04 10:11:00   
...                    ...                ...                 ...   
18543  5805028000056889209      Ulysses Adams 2024-06-21 12:11:00   
18544  5805028000056889351           Eva Kent 2024-06-21 13:32:00   
18545  5805028000056892018           Eva Kent 2024-06-21 10:21:00   
18546  5805028000056892055       Yara Edwards 2024-06-21 10:22:00   
18547  5805028000056907001           Ben Hall 2024-06-21 10:56:00   

            Modified Time  
0     2023-12-22 13:34:00  
1     2024-05-21 10:23:00  
2     2023-12-21 13:17:00  
3     2023-12-29 15:20:00  
4     2024-04-16 16:14:00  
...                   ...  
18543 2024-06-21 14:11:00  
18544 2024-06-21 15:32:00  
18545 2024-06-21 12:21:00  
18546 2024-06-21 12:23:00  
18547 2024-06-21 12:56:00  

[18510 rows x 4 columns]

In [ ]:
contacts_df.to_excel("Contacts (Result).xlsx", index=False)

============================================================================================



##✨ **Data Cleaning `calls_df`**

In [ ]:
df_path = '/content/Calls (Done).xlsx'
calls_df = load_data(df_path, 'xlsx', dtype={'Id': str, "CONTACTID": str})
calls_df.head()

Id   Call Start Time Call Owner Name            CONTACTID  \
0  5805028000000805001  30.06.2023 08:43        John Doe                  NaN   
1  5805028000000768006  30.06.2023 08:46        John Doe                  NaN   
2  5805028000000764027  30.06.2023 08:59        John Doe                  NaN   
3  5805028000000787003  30.06.2023 09:20        John Doe  5805028000000645014   
4  5805028000000768019  30.06.2023 09:30        John Doe  5805028000000645014   

  Call Type  Call Duration (in seconds)       Call Status  Dialled Number  \
0   Inbound                       171.0          Received             NaN   
1  Outbound                        28.0  Attended Dialled             NaN   
2  Outbound                        24.0  Attended Dialled             NaN   
3  Outbound                         6.0  Attended Dialled             NaN   
4  Outbound                        11.0  Attended Dialled             NaN   

  Outgoing Call Status  Scheduled in CRM  Tag  
0                  NaN               NaN  NaN  
1            Completed               0.0  NaN  
2            Completed               0.0  NaN  
3            Completed               0.0  NaN  
4            Completed               0.0  NaN

In [ ]:
calls_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 95874 entries, 0 to 95873
Data columns (total 11 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   Id                          95874 non-null  object 
 1   Call Start Time             95874 non-null  object 
 2   Call Owner Name             95874 non-null  object 
 3   CONTACTID                   91941 non-null  object 
 4   Call Type                   95874 non-null  object 
 5   Call Duration (in seconds)  95791 non-null  float64
 6   Call Status                 95874 non-null  object 
 7   Dialled Number              0 non-null      float64
 8   Outgoing Call Status        86875 non-null  object 
 9   Scheduled in CRM            86875 non-null  float64
 10  Tag                         0 non-null      float64
dtypes: float64(4), object(7)
memory usage: 8.0+ MB


In [ ]:
calls_df.describe()

Call Duration (in seconds)  Dialled Number  Scheduled in CRM  Tag
count                95791.000000             0.0      86875.000000  0.0
mean                   164.977263             NaN          0.001635  NaN
std                    401.410826             NaN          0.040397  NaN
min                      0.000000             NaN          0.000000  NaN
25%                      4.000000             NaN          0.000000  NaN
50%                      8.000000             NaN          0.000000  NaN
75%                     98.000000             NaN          0.000000  NaN
max                   7625.000000             NaN          1.000000  NaN

---
## Removing duplicates
---

In [ ]:
calls_df.duplicated(subset=calls_df.columns[1:]).sum()

3257

In [ ]:
calls_df[calls_df.duplicated(keep=False, subset=calls_df.columns[1:])]

Id   Call Start Time Call Owner Name  \
34     5805028000001140014  06.07.2023 17:15   Alice Johnson   
35     5805028000001167001  06.07.2023 17:15   Alice Johnson   
101    5805028000001372054  08.07.2023 16:43        John Doe   
102    5805028000001348077  08.07.2023 16:43        John Doe   
254    5805028000001568042  12.07.2023 19:23      Jane Smith   
...                    ...               ...             ...   
95804  5805028000056832311  21.06.2024 14:17    Yara Edwards   
95833  5805028000056845313  21.06.2024 14:47   Ulysses Adams   
95834  5805028000056873560  21.06.2024 14:47   Ulysses Adams   
95838  5805028000056834447  21.06.2024 14:55        John Doe   
95839  5805028000056892435  21.06.2024 14:55        John Doe   

                 CONTACTID Call Type  Call Duration (in seconds)  \
34     5805028000001129001  Outbound                         0.0   
35     5805028000001129001  Outbound                         0.0   
101                    NaN    Missed                         0.0   
102                    NaN    Missed                         0.0   
254    5805028000001552025  Outbound                         0.0   
...                    ...       ...                         ...   
95804                  NaN  Outbound                         8.0   
95833  5805028000026041053  Outbound                         0.0   
95834  5805028000026041053  Outbound                         0.0   
95838                  NaN    Missed                         0.0   
95839                  NaN    Missed                         0.0   

              Call Status  Dialled Number Outgoing Call Status  \
34     Unattended Dialled             NaN            Completed   
35     Unattended Dialled             NaN            Completed   
101                Missed             NaN                  NaN   
102                Missed             NaN                  NaN   
254    Unattended Dialled             NaN            Completed   
...                   ...             ...                  ...   
95804    Attended Dialled             NaN            Completed   
95833  Unattended Dialled             NaN            Completed   
95834  Unattended Dialled             NaN            Completed   
95838              Missed             NaN                  NaN   
95839              Missed             NaN                  NaN   

       Scheduled in CRM  Tag  
34                  0.0  NaN  
35                  0.0  NaN  
101                 NaN  NaN  
102                 NaN  NaN  
254                 0.0  NaN  
...                 ...  ...  
95804               0.0  NaN  
95833               0.0  NaN  
95834               0.0  NaN  
95838               NaN  NaN  
95839               NaN  NaN  

[6382 rows x 11 columns]

In [ ]:
calls_df.drop_duplicates(subset=calls_df.columns[1:], inplace=True)

calls_df[calls_df.duplicated(subset=calls_df.columns[1:])].shape

(0, 11)

---
## Processing Missing Values and Data Type Conversions
---

In [ ]:
calls_df.isna().sum()

Id                                0
Call Start Time                   0
Call Owner Name                   0
CONTACTID                      3802
Call Type                         0
Call Duration (in seconds)       79
Call Status                       0
Dialled Number                92617
Outgoing Call Status           8813
Scheduled in CRM               8813
Tag                           92617
dtype: int64

#### ☑ columns'Dialled Number' and 'Tag' - remove columns as all values ​​are NaN



In [ ]:
calls_df = calls_df.drop(columns=['Dialled Number', 'Tag'])

In [ ]:
calls_df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 92617 entries, 0 to 95873
Data columns (total 9 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   Id                          92617 non-null  object 
 1   Call Start Time             92617 non-null  object 
 2   Call Owner Name             92617 non-null  object 
 3   CONTACTID                   88815 non-null  object 
 4   Call Type                   92617 non-null  object 
 5   Call Duration (in seconds)  92538 non-null  float64
 6   Call Status                 92617 non-null  object 
 7   Outgoing Call Status        83804 non-null  object 
 8   Scheduled in CRM            83804 non-null  float64
dtypes: float64(2), object(7)
memory usage: 7.1+ MB


In [ ]:
calls_df['Call Start Time'] = pd.to_datetime(calls_df['Call Start Time'], format='%d.%m.%Y %H:%M', dayfirst=True, errors='raise')

In [ ]:
category_columns = ['Call Owner Name', 'Call Type', 'Call Status', 'Outgoing Call Status', 'Scheduled in CRM']

for col in category_columns:
    calls_df[col] = calls_df[col].astype('category')

for col in ['Outgoing Call Status', 'Scheduled in CRM']:
    if 'Unknown' not in calls_df[col].cat.categories:
        calls_df[col] = calls_df[col].cat.add_categories(['Unknown'])
    calls_df[col] = calls_df[col].fillna('Unknown')

In [ ]:
calls_df["CONTACTID"] = calls_df["CONTACTID"].fillna("Unknown")

---
## Results Output
---

In [ ]:
calls_df.head()

Id     Call Start Time Call Owner Name  \
0  5805028000000805001 2023-06-30 08:43:00        John Doe   
1  5805028000000768006 2023-06-30 08:46:00        John Doe   
2  5805028000000764027 2023-06-30 08:59:00        John Doe   
3  5805028000000787003 2023-06-30 09:20:00        John Doe   
4  5805028000000768019 2023-06-30 09:30:00        John Doe   

             CONTACTID Call Type  Call Duration (in seconds)  \
0              Unknown   Inbound                       171.0   
1              Unknown  Outbound                        28.0   
2              Unknown  Outbound                        24.0   
3  5805028000000645014  Outbound                         6.0   
4  5805028000000645014  Outbound                        11.0   

        Call Status Outgoing Call Status Scheduled in CRM  
0          Received              Unknown          Unknown  
1  Attended Dialled            Completed              0.0  
2  Attended Dialled            Completed              0.0  
3  Attended Dialled            Completed              0.0  
4  Attended Dialled            Completed              0.0

In [ ]:
calls_df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 92617 entries, 0 to 95873
Data columns (total 9 columns):
 #   Column                      Non-Null Count  Dtype         
---  ------                      --------------  -----         
 0   Id                          92617 non-null  object        
 1   Call Start Time             92617 non-null  datetime64[ns]
 2   Call Owner Name             92617 non-null  category      
 3   CONTACTID                   92617 non-null  object        
 4   Call Type                   92617 non-null  category      
 5   Call Duration (in seconds)  92538 non-null  float64       
 6   Call Status                 92617 non-null  category      
 7   Outgoing Call Status        92617 non-null  category      
 8   Scheduled in CRM            92617 non-null  category      
dtypes: category(5), datetime64[ns](1), float64(1), object(2)
memory usage: 4.0+ MB


In [ ]:
calls_df

Id     Call Start Time Call Owner Name  \
0      5805028000000805001 2023-06-30 08:43:00        John Doe   
1      5805028000000768006 2023-06-30 08:46:00        John Doe   
2      5805028000000764027 2023-06-30 08:59:00        John Doe   
3      5805028000000787003 2023-06-30 09:20:00        John Doe   
4      5805028000000768019 2023-06-30 09:30:00        John Doe   
...                    ...                 ...             ...   
95869  5805028000056889515 2024-06-21 15:30:00   Ulysses Adams   
95870  5805028000056875317 2024-06-21 15:30:00   Victor Barnes   
95871  5805028000056832495 2024-06-21 15:30:00    Kevin Parker   
95872  5805028000056893619 2024-06-21 15:30:00   Victor Barnes   
95873  5805028000056893631 2024-06-21 15:31:00        John Doe   

                 CONTACTID Call Type  Call Duration (in seconds)  \
0                  Unknown   Inbound                       171.0   
1                  Unknown  Outbound                        28.0   
2                  Unknown  Outbound                        24.0   
3      5805028000000645014  Outbound                         6.0   
4      5805028000000645014  Outbound                        11.0   
...                    ...       ...                         ...   
95869  5805028000056564231  Outbound                         6.0   
95870  5805028000054867023  Outbound                         8.0   
95871  5805028000010617278  Outbound                         5.0   
95872  5805028000056839048  Outbound                         0.0   
95873              Unknown    Missed                         0.0   

              Call Status Outgoing Call Status Scheduled in CRM  
0                Received              Unknown          Unknown  
1        Attended Dialled            Completed              0.0  
2        Attended Dialled            Completed              0.0  
3        Attended Dialled            Completed              0.0  
4        Attended Dialled            Completed              0.0  
...                   ...                  ...              ...  
95869    Attended Dialled            Completed              0.0  
95870    Attended Dialled            Completed              0.0  
95871    Attended Dialled            Completed              0.0  
95872  Unattended Dialled            Completed              0.0  
95873              Missed              Unknown          Unknown  

[92617 rows x 9 columns]

In [ ]:
calls_df.to_excel("Calls (Result).xlsx", index=False)

============================================================================================

##✨ **Data Cleaning `spend_df`**




In [ ]:
df_path = '/content/Spend (Done).xlsx'
spend_df = load_data(df_path, 'xlsx')
spend_df.head()

Date        Source               Campaign  Impressions  Spend  Clicks  \
0 2023-07-03    Google Ads         gen_analyst_DE            6   0.00       0   
1 2023-07-03    Google Ads  performancemax_eng_DE            4   0.01       1   
2 2023-07-03  Facebook Ads                    NaN            0   0.00       0   
3 2023-07-03    Google Ads                    NaN            0   0.00       0   
4 2023-07-03           CRM                    NaN            0   0.00       0   

  AdGroup   Ad  
0     NaN  NaN  
1     NaN  NaN  
2     NaN  NaN  
3     NaN  NaN  
4     NaN  NaN

In [ ]:
spend_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 20779 entries, 0 to 20778
Data columns (total 8 columns):
 #   Column       Non-Null Count  Dtype         
---  ------       --------------  -----         
 0   Date         20779 non-null  datetime64[ns]
 1   Source       20779 non-null  object        
 2   Campaign     14785 non-null  object        
 3   Impressions  20779 non-null  int64         
 4   Spend        20779 non-null  float64       
 5   Clicks       20779 non-null  int64         
 6   AdGroup      13951 non-null  object        
 7   Ad           13951 non-null  object        
dtypes: datetime64[ns](1), float64(1), int64(2), object(4)
memory usage: 1.3+ MB


---
## Data Type Conversions
---

In [ ]:
spend_df['Source'] = spend_df['Source'].astype('category')
spend_df['Campaign'] = spend_df['Campaign'].fillna('Unknown').astype('category')

### Removing duplicates

In [ ]:
spend_df.duplicated(subset=spend_df.columns[1:]).sum()

5722

In [ ]:
spend_df[spend_df.duplicated(keep=False, subset=spend_df.columns[1:])]

Date          Source Campaign  Impressions  Spend  Clicks AdGroup  \
2     2023-07-03    Facebook Ads  Unknown            0    0.0       0     NaN   
3     2023-07-03      Google Ads  Unknown            0    0.0       0     NaN   
4     2023-07-03             CRM  Unknown            0    0.0       0     NaN   
7     2023-07-03        Bloggers  Unknown            0    0.0       0     NaN   
8     2023-07-03     Youtube Ads  Unknown            0    0.0       0     NaN   
...          ...             ...      ...          ...    ...     ...     ...   
20764 2024-06-21  Telegram posts  Unknown            0    0.0       0     NaN   
20770 2024-06-21         Organic  Unknown            0    0.0       0     NaN   
20773 2024-06-21         Organic  Unknown            0    0.0       0     NaN   
20776 2024-06-21     Partnership  Unknown            0    0.0       0     NaN   
20777 2024-06-21      Tiktok Ads  Unknown            0    0.0       0     NaN   

        Ad  
2      NaN  
3      NaN  
4      NaN  
7      NaN  
8      NaN  
...    ...  
20764  NaN  
20770  NaN  
20773  NaN  
20776  NaN  
20777  NaN  

[6367 rows x 8 columns]

In [ ]:
spend_df.drop_duplicates(subset=spend_df.columns[1:], inplace=True)

spend_df[spend_df.duplicated(subset=spend_df.columns[1:])].shape

(0, 8)

---
### Processing Missing Values
---

In [ ]:
spend_df.isna().sum()

Date              0
Source            0
Campaign          0
Impressions       0
Spend             0
Clicks            0
AdGroup        2087
Ad             2087
dtype: int64

In [ ]:
(spend_df.isna().sum() + (spend_df == 0).sum())

Date              0
Source            0
Campaign          0
Impressions     533
Spend          1153
Clicks         4024
AdGroup        2087
Ad             2087
dtype: int64

In [ ]:
spend_df[spend_df['Impressions']== 0][['Spend', 'Campaign', 'Impressions', 'Clicks']]

Spend                 Campaign  Impressions  Clicks
2        0.00                  Unknown            0       0
3        0.00                  Unknown            0       0
4        0.00                  Unknown            0       0
7        0.00                  Unknown            0       0
8        0.00                  Unknown            0       0
...       ...                      ...          ...     ...
20642    0.00  08.04.24wide_webinar_DE            0       0
20643    0.00  08.04.24wide_webinar_DE            0       0
20712    4.25                  Unknown            0       0
20755  200.00                  Unknown            0       0
20763   74.25                  Unknown            0       0

[533 rows x 4 columns]

In [ ]:
spend_df[spend_df['Impressions']== 0][['Spend', 'Impressions', 'Clicks']].sum()

Spend           7238.5
Impressions        0.0
Clicks         44486.0
dtype: float64

In [ ]:
spend_df[(spend_df[['Spend', 'Impressions', 'Clicks']] == 0).all(axis=1)][['Spend', 'Impressions', 'Clicks']]

Spend  Impressions  Clicks
2        0.0            0       0
3        0.0            0       0
4        0.0            0       0
7        0.0            0       0
8        0.0            0       0
...      ...          ...     ...
20572    0.0            0       0
20610    0.0            0       0
20616    0.0            0       0
20642    0.0            0       0
20643    0.0            0       0

[90 rows x 3 columns]

#### ☑ We remove rows where Spend, Impressions, and Clicks are all zero, as they do not provide useful information and may distort data analysis.

In [ ]:
spend_df = spend_df[~(spend_df[['Spend', 'Impressions', 'Clicks']] == 0).all(axis=1)]

In [ ]:
spend_df[(spend_df[['Spend', 'Impressions', 'Clicks']] == 0).all(axis=1)][['Spend', 'Impressions', 'Clicks']]

Empty DataFrame
Columns: [Spend, Impressions, Clicks]
Index: []

In [ ]:
spend_df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 14967 entries, 0 to 20778
Data columns (total 8 columns):
 #   Column       Non-Null Count  Dtype         
---  ------       --------------  -----         
 0   Date         14967 non-null  datetime64[ns]
 1   Source       14967 non-null  category      
 2   Campaign     14967 non-null  category      
 3   Impressions  14967 non-null  int64         
 4   Spend        14967 non-null  float64       
 5   Clicks       14967 non-null  int64         
 6   AdGroup      12894 non-null  object        
 7   Ad           12894 non-null  object        
dtypes: category(2), datetime64[ns](1), float64(1), int64(2), object(2)
memory usage: 850.9+ KB


---
### Results Output
---

In [ ]:
spend_df.head()

Date        Source               Campaign  Impressions  Spend  Clicks  \
0 2023-07-03    Google Ads         gen_analyst_DE            6   0.00       0   
1 2023-07-03    Google Ads  performancemax_eng_DE            4   0.01       1   
5 2023-07-03  Facebook Ads          03.07.23women          187   3.30       6   
6 2023-07-03  Facebook Ads          03.07.23women            4   0.02       1   
9 2023-07-03  Facebook Ads        02.07.23wide_DE           61   0.58       0   

  AdGroup   Ad  
0     NaN  NaN  
1     NaN  NaN  
5   women   b3  
6   women   b1  
9    wide   b4

In [ ]:
spend_df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 14967 entries, 0 to 20778
Data columns (total 8 columns):
 #   Column       Non-Null Count  Dtype         
---  ------       --------------  -----         
 0   Date         14967 non-null  datetime64[ns]
 1   Source       14967 non-null  category      
 2   Campaign     14967 non-null  category      
 3   Impressions  14967 non-null  int64         
 4   Spend        14967 non-null  float64       
 5   Clicks       14967 non-null  int64         
 6   AdGroup      12894 non-null  object        
 7   Ad           12894 non-null  object        
dtypes: category(2), datetime64[ns](1), float64(1), int64(2), object(2)
memory usage: 850.9+ KB


In [ ]:
spend_df

Date        Source                  Campaign  Impressions  Spend  \
0     2023-07-03    Google Ads            gen_analyst_DE            6   0.00   
1     2023-07-03    Google Ads     performancemax_eng_DE            4   0.01   
5     2023-07-03  Facebook Ads             03.07.23women          187   3.30   
6     2023-07-03  Facebook Ads             03.07.23women            4   0.02   
9     2023-07-03  Facebook Ads           02.07.23wide_DE           61   0.58   
...          ...           ...                       ...          ...    ...   
20771 2024-06-21    Tiktok Ads         22.05.2024wide_DE            7   0.03   
20772 2024-06-21   Youtube Ads         youtube_shorts_DE           90   0.51   
20774 2024-06-21  Facebook Ads           17.03.24wide_AT            7   0.07   
20775 2024-06-21    Tiktok Ads         12.07.2023wide_DE           61   0.16   
20778 2024-06-21  Facebook Ads  04.07.23recentlymoved_DE            7   0.05   

       Clicks        AdGroup                     Ad  
0           0            NaN                    NaN  
1           1            NaN                    NaN  
5           6          women                     b3  
6           1          women                     b1  
9           0           wide                     b4  
...       ...            ...                    ...  
20771       0           wide     bloggersvideo18com  
20772       2     Com_august     bloggersvideo3june  
20774       0           wide  bloggersvideo16com_at  
20775       0           wide     bloggersvideo14com  
20778       1  recentlymoved     bloggersvideo14com  

[14967 rows x 8 columns]

In [ ]:
spend_df.to_excel("Spend (Result).xlsx", index=False)

In [ ]:
contacts_df.to_excel("Contacts (Result).xlsx", index=False)

============================================================================================



##✨ **Data Cleaning `deals_df`**

In [ ]:
path = '/content/Deals (Done).xlsx'
deals_df = load_data(path, 'xlsx', dtype={'Id': str, 'Contact Name': str})
deals_df.head()

Id Deal Owner Name Closing Date            Quality  \
0  5805028000056864695        Ben Hall          NaN                NaN   
1  5805028000056859489   Ulysses Adams          NaN                NaN   
2  5805028000056832357   Ulysses Adams   21.06.2024     D - Non Target   
3  5805028000056824246        Eva Kent   21.06.2024  E - Non Qualified   
4  5805028000056873292        Ben Hall   21.06.2024     D - Non Target   

      Stage     Lost Reason       Page                  Campaign       SLA  \
0  New Lead             NaN  /eng/test             03.07.23women       NaN   
1  New Lead             NaN    /at-eng                       NaN       NaN   
2      Lost      Non target    /at-eng                engwien_AT  00:26:43   
3      Lost  Invalid number       /eng  04.07.23recentlymoved_DE  01:00:04   
4      Lost      Non target       /eng              discovery_DE  00:53:12   

              Content  ...        Product Education Type      Created Time  \
0                 v16  ...            NaN            NaN  21.06.2024 15:30   
1                 NaN  ...  Web Developer        Morning  21.06.2024 15:23   
2               b1-at  ...            NaN            NaN  21.06.2024 14:45   
3  bloggersvideo14com  ...            NaN            NaN  21.06.2024 13:32   
4             website  ...            NaN            NaN  21.06.2024 13:21   

  Course duration Months of study Initial Amount Paid  Offer Total Amount  \
0             NaN             NaN                 NaN                 NaN   
1             6.0             NaN                   0                2000   
2             NaN             NaN                 NaN                 NaN   
3             NaN             NaN                 NaN                 NaN   
4             NaN             NaN                 NaN                 NaN   

          Contact Name City Level of Deutsch  
0  5805028000056849495  NaN              NaN  
1  5805028000056834471  NaN              NaN  
2  5805028000056854421  NaN              NaN  
3  5805028000056889351  NaN              NaN  
4  5805028000056876176  NaN              NaN  

[5 rows x 23 columns]

In [ ]:
deals_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21595 entries, 0 to 21594
Data columns (total 23 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   Id                   21593 non-null  object 
 1   Deal Owner Name      21564 non-null  object 
 2   Closing Date         14645 non-null  object 
 3   Quality              19340 non-null  object 
 4   Stage                21593 non-null  object 
 5   Lost Reason          16124 non-null  object 
 6   Page                 21593 non-null  object 
 7   Campaign             16067 non-null  object 
 8   SLA                  15533 non-null  object 
 9   Content              14147 non-null  object 
 10  Term                 12454 non-null  object 
 11  Source               21593 non-null  object 
 12  Payment Type         496 non-null    object 
 13  Product              3592 non-null   object 
 14  Education Type       3300 non-null   object 
 15  Created Time         21593 non-null 

In [ ]:
deals_df.tail()

Id Deal Owner Name Closing Date            Quality  \
21590  5805028000000945016      Jane Smith   29.08.2023           A - High   
21591  5805028000000927004       Bob Brown   09.07.2023     D - Non Target   
21592  5805028000000922001       Bob Brown   03.07.2023  E - Non Qualified   
21593                  NaN             NaN          NaN                NaN   
21594                  NaN             NaN          NaN                NaN   

      Stage             Lost Reason                   Page         Campaign  \
21590  Lost        Changed Decision  eng/digital-marketing  02.07.23wide_DE   
21591  Lost  Does not speak English  eng/digital-marketing    03.07.23women   
21592  Lost                 Refugee                      /              NaN   
21593   NaN                     NaN                    NaN              NaN   
21594   NaN                     NaN                    NaN              NaN   

                     SLA Content  ... Product Education Type  \
21590  56 days, 19:01:59      b3  ...     NaN            NaN   
21591                NaN      b3  ...     NaN            NaN   
21592   4 days, 22:47:14     NaN  ...     NaN            NaN   
21593                NaN     NaN  ...     NaN            NaN   
21594                NaN     NaN  ...     NaN          #REF!   

           Created Time Course duration Months of study Initial Amount Paid  \
21590  03.07.2023 20:39             NaN             NaN                 NaN   
21591  03.07.2023 20:17             NaN             NaN                 NaN   
21592  03.07.2023 17:03             NaN             NaN                   0   
21593               NaN             NaN             NaN                 NaN   
21594               NaN             NaN             NaN                 NaN   

       Offer Total Amount         Contact Name City Level of Deutsch  
21590                 NaN  5805028000000968001  NaN              NaN  
21591                 NaN  5805028000000961001  NaN              NaN  
21592                   0  5805028000001009140  NaN              NaN  
21593                 NaN                  NaN  NaN              NaN  
21594                 NaN                  NaN  NaN              NaN  

[5 rows x 23 columns]

#### ☑ Removed rows with missing values in the 'Id' column to ensure data integrity and avoid errors in analysis

In [ ]:
deals_df.dropna(subset=["Id"], inplace=True)

In [ ]:
deals_df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 21593 entries, 0 to 21592
Data columns (total 23 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   Id                   21593 non-null  object 
 1   Deal Owner Name      21564 non-null  object 
 2   Closing Date         14645 non-null  object 
 3   Quality              19340 non-null  object 
 4   Stage                21593 non-null  object 
 5   Lost Reason          16124 non-null  object 
 6   Page                 21593 non-null  object 
 7   Campaign             16067 non-null  object 
 8   SLA                  15533 non-null  object 
 9   Content              14147 non-null  object 
 10  Term                 12454 non-null  object 
 11  Source               21593 non-null  object 
 12  Payment Type         496 non-null    object 
 13  Product              3592 non-null   object 
 14  Education Type       3299 non-null   object 
 15  Created Time         21593 non-null  obje

---
## Data Type Conversions
---

In [ ]:
print(deals_df['Initial Amount Paid'].unique())
print(deals_df['Offer Total Amount'].unique())

[nan 0 1000 '€ 3.500,00' 500 100 4500 300 200 2000 11000 4000 3000 3500
 11500 1200 1500 1 5000 600 700 350 9 400 450]
[nan 2000 9000 11000 3500 4500 '€ 2.900,00' 6500 4000 3000 10000 2500 5000
 11500 1 1000 1200 0 1500 '€ 11398,00' 11111 6000]


In [ ]:
def clean_currency_col(df, col_name):
    df[col_name] = (df[col_name]
                       .replace(r'[€]', '', regex=True)
                       .replace(r'\s+', '', regex=True)
                       .replace(r'\.', '', regex=True)
                       .replace(r',', '.', regex=True)
                       .astype(float))
    return df

In [ ]:
deals_df = clean_currency_col(deals_df, 'Initial Amount Paid')
deals_df = clean_currency_col(deals_df, 'Offer Total Amount')

In [ ]:
def convert_to_seconds(x):
    if pd.isna(x):
        return np.nan
    elif isinstance(x, datetime.time):
        return x.hour * 3600 + x.minute * 60 + x.second
    elif isinstance(x, datetime.timedelta):
        return x.total_seconds()

deals_df['SLA Seconds'] = deals_df["SLA"].apply(convert_to_seconds)
deals_df['SLA Seconds'].value_counts()

SLA Seconds
611.0       6
669.0       6
646.0       6
697.0       5
1028.0      5
           ..
25870.0     1
54028.0     1
43761.0     1
40467.0     1
427634.0    1
Name: count, Length: 13357, dtype: int64

In [ ]:
deals_df.drop(columns=['SLA'], inplace=True)

In [ ]:
deals_df['Closing Date'] = pd.to_datetime(deals_df['Closing Date'], format='%d.%m.%Y', dayfirst=True, errors='raise')
deals_df['Created Time'] = pd.to_datetime(deals_df['Created Time'], format='%d.%m.%Y %H:%M', dayfirst=True, errors='raise')
deals_df[['Course duration', 'Months of study']] = deals_df[['Course duration',  'Months of study']].astype('Int8')

---
## Processing Missing Values and Data Standardization
---

#### ☑ We fill  column with the most common city for each 'Contact Name', imputing missing or inconsistent city values based on the mode for each contact. This helps standardize the data and improve its reliability for analysis.

In [ ]:
mode_values = deals_df.groupby(
    'Contact Name')['City'].agg(lambda x: x.mode()[0] if not x.mode()\
                           .empty else None)
deals_df['City'] = deals_df['Contact Name'].map(mode_values)

#### ☑ We fill  column with the most common city for each 'Contact Name', imputing missing or inconsistent 'Level of Deutsch' values based on the mode for each contact.

In [ ]:
mode_values = deals_df.groupby(
    'Contact Name')['Level of Deutsch']\
    .agg(lambda x: x.mode()[0] if not x.mode().empty else None)
deals_df['Level of Deutsch'] = deals_df['Contact Name'].map(mode_values)

#### ☑ We standardize entries in the 'Level of Deutsch' column to unify the data

In [ ]:
deals_df['Level of Deutsch'].unique()

array([None, 'б1', 'в1', 'A2', 'в2', 'b1', 'В1', 'B1', 'в1-в2', 'А2-В1',
       'А2 ( Б1 в июне)', '-', 'B2', 'C2', 'с1', 'Б1', 'Б2',
       'b1 ждет результаты', 'а2', 'Б1( может будет)', 'а1', 'С1',
       'a2 (b1 экзамен 15 июня)', 'а0', 'б2', 'А2',
       'B1 будет в феврале 2025', 'Detmold, Paulinenstraße 95, 32756',
       'Сам оценивает на B2, 13 лет живет в Германии', 'В1-В2',
       'Б1 ( ждет Б2)',
       'lэкзамен - 6 июля на В1. курсы вечером (но уверенно говорит на B1)',
       'Гражданка Германии уже год в Германии Учит немецкий и в сентябре b1 через гос-во проходит, а не через ДЖЦ, вечером учится 3 р в неделю с 18 до 21',
       'B1 в процессе обучения',
       'ЯЗ: нем В1 был экз 03.05 повтор и сейчас ждет результаты. Технический англ был. А1 сейчас. ОБР: 2 во информационные и комп сети - инженер системоте',
       'В1 в сентябре', 'Нет', 0, 'Ждем B1', 'А1 сертиф, но по факту А2',
       'a2', 'Пока А2, сдает 17 05 B1', 'окончание 13.06 курса на b1',
       'A1', 'b2', 

In [ ]:
replace_dict = {
    'в1': 'B1', 'б1': 'B1', 'b1': 'B1', 'B1': 'B1', 'в1-в2': 'B1-B2', 'B2': 'B2',
    'C2': 'C2', 'с1': 'C1', 'Б1': 'B1', 'а2': 'A2', 'а1': 'A1', 'а0': 'A0',
    'б2': 'B2', 'Б2': 'B2', 'В1': 'B1', 'А2': 'A2',
    'B1 будет в феврале 2025': 'B1',
    'Detmold, Paulinenstraße 95, 32756': 'Неизвестно',
    'Сам оценивает на B2, 13 лет живет в Германии': 'B2',
    'в2': 'B2', 'В1-В2': 'B1-B2', 'Б1 ( ждет Б2)': 'B1-B2', 'А2-В1': 'A2-B1',
    'lэкзамен - 6 июля на В1. курсы вечером (но уверенно говорит на B1)': 'B1',
    'Гражданка Германии уже год в Германии Учит немецкий и в сентябре b1 через гос-во проходит, а не через ДЖЦ, вечером учится 3 р в неделю с 18 до 21': 'B1',
    '-': 'A0', 'А2 ( Б1 в июне)': 'A2-B1', 'B1 в процессе обучения': 'B1',
    'ЯЗ: нем В1 был экз 03.05 повтор и сейчас ждет результаты. Технический англ был. А1 сейчас. ОБР: 2 во информационные и комп сети - инженер системоте': 'B1',
    'В1 в сентябре': 'B1', 'Нет': 'Неизвестно', 'С1': 'C1', 0: 'A0',
    'Ждем B1': 'B1', 'А1 сертиф, но по факту А2': 'A2', 'a2': 'A2',
    'Пока А2, сдает 17 05 B1': 'A2-B1', 'окончание 13.06 курса на b1': 'B1',
    'A1': 'A1', 'b2': 'B2', 'Thorn-Prikker-Str. 30, Hagen, 58093': 'Неизвестно',
    'В2': 'B2', 'нулевой уровень, только пошел на курсы.': 'A0', 'ая в1': 'B1',
    'Ждет результат по B1': 'B1', 'А2( ждет итоги Б!)': 'A2-B1',
    'b1 экзамен будет 12 апреля': 'B1', 'b1 (b2 ждет серт)': 'B1-B2', 'С2': 'C2',
    'ждем B1': 'B1', 'Paderborn 33102, Schwabenweg 10': 'Неизвестно',
    'b1 (B2 ждет серт)': 'B1-B2', 'Ждем B1 со дня на день': 'B1',
    'Б2 ( учит С1)': 'B2-C1', 'B1 еще нет результата': 'B1',
    '31.05.2024': 'Неизвестно', 'Lichtenfelser Straße 25, Untersiemau 96253': 'Неизвестно',
    'Учиться до сентября на B1': 'B1', 'b1 9ждет экзамен)': 'B1',
    'b1+': 'B1', 'гражданка': 'B2',
    'b1 (ждет результат)': 'B1', 'Б1 (учит Б2)': 'B1-B2',
    'б2+': 'B2', 'Гражданин': 'B2', '25 лет живет в Германии': 'C1',
    'С1 -ая , Ня -а1': 'C1', 'Ждем результат по B1': 'B1',
    'b1 (b2 в июле экзамен)': 'B1-B2', 'Ждет со дня на день В1': 'B1',
    'А2 (В1 с 3 раза не сдала, бератер видела наши доки)': 'A2-B1',
    'b1 (ждет результаты)': 'B1',
    'А2 ( повтор на Б1)': 'A2-B1', 'B1, сдает B2 в апреле': 'B1-B2',
    'ждет сертификат B1': 'B1',
    'Б2( 16.02 экзамен С1)': 'B2-C1', 'А1-А2': 'A1-A2',
    'b1 ждет серт на днях на руки': 'B1',
    'b1 24 февраля экзамен, англ b2': 'B1-B2', 'А2 ( скоро екзамен)': 'A2',
    'B1 (ждет результаты В2)': 'B1-B2',
    'b1 (b2 15 марта экзамен)': 'B1-B2', 'b2 (с1 экзамен 16 февраля)': 'B2-C1',
    'Б1 ( ждет итог Б2)': 'B1-B2', 'не сдавал, но гражданин': 'B2',
    'Нет сертификатов, но есть С1 англ, неоконченное высшее в ИТ (и еще одно высшее юридическое) , очень хочет в ИТ, сильно замотивирована именно н': 'Неизвестно',
    'А2, в процессе Б1': 'A2-B1', 'A0': 'A0', 'А2(Б1 в марте экз)': 'A2-B1',
    'учит A2': 'A2', 'Б1 ( проходит Б2)': 'B1-B2', 'Б1 ( ждет итог )': 'B1-B2',
    'НЯ - В1, АЯ - В1': 'B1', 'б1 (ждет рез-тат)': 'B1',
    'А2(ждет итоги Б1)': 'A2-B1', 'в1-ня , в1-ая': 'B1',
    'ня-0, но англ B2+': 'A0', 'В': 'B1', 'будет B1 в июне': 'B1',
    'А2( включили нем в ангебот)': 'A2', 'а2-в1': 'A2-B1', 'в2-с1': 'B2-C1',
    'курс А2-В1 - сдача в июле, но вечерняя смена инт курсов, настроен получить гутшайн уже сейчас.': 'A2-B1',
    'B1 (B2 должна до конца февраля получить)': 'B1-B2',
    'b1 (b2 экзамен 6 февраля)': 'B1-B2',
    'A1-A2': 'A1-A2', 'Б1( может будет)': 'B1', 'А2 ( в процессе Б1)': 'A2-B1',
    'b1 ждет результаты': 'B1',
    'b1 ждет экзамен в феврале': 'B1', 'В1, может уже В2?': 'B1-B2',
    'A2 (идет доучивать В1 - 300 часов; предположительно до августа)': 'A2-B1',
    'не учил': 'A0', 'Без 5 минут B1 (ждет результаты экзамена)': 'B1',
    'а1-а2 , ая свободный': 'A1-A2', 'b2-c2': 'B2',
    'а2, англ B1': 'A2', 'А1': 'A1', 'А2 нем -В2 англ': 'A2',
    'Проходит сейчас B1': 'B1',
    'Ждет результат по B1 в феврале': 'B1', 'Проходит сейчас повторно B1': 'B1',
    'b1 экзамен в феврале': 'B1',
    'Учиться на B1 во вторую смену, в первую хочет получить одобрение на обучение у нас': 'B1',
    'Б10Б2': 'B2', 'Б1?': 'B1', 'B1 есть, ждем B2 в конце месяца': 'B1-B2',
    'B1-B2': 'B1-B2', '?': 'Неизвестно', 'b1 экзамен 26 января': 'B1', 'А0': 'A0',
    'а2 (б1 в сер января)': 'A2-B1', 'f2': 'A2', 'Учиться на B1': 'B1',
    'Сдала экзамен на B1, ждет в начале февраля результат': 'B1',
    'Сдавал 8 12 на B1 - ждет результат. 3 01 - аплейт - получил B1!': 'B1',
    'Б1-Б2': 'B1-B2',     'б1 (до июля на В2)': 'B1-B2',
    'А2 ( Б1 март )': 'A2-B1', 'А2 (весной - еще 300 часов В1)': 'A2-B1',
    'В январе будут результаты по экзамену на B1': 'B1',
    'б2 (с1 ждет рез-тат)': 'B2-C1', 'ня-0, ая-B1': 'A0', 'А2-Б1': 'A2-B1',
    'B1 (почти, не сдала чуть) + англ В1': 'B1', 'в1 ждем результаты': 'B1',
    'А2 ( хочет просить совмещать)': 'A2', 'B1 (ждет результаты)': 'B1',
    'А2+': 'A2', 'а2 (сдавала экз В1, но не сдала похоже)': 'A2',
    'в1, идет на в2': 'B1-B2', 'b2-c1': 'B2-C1', 'C1': 'C1', 'b1-b2': 'B1-B2',
    'не учила ( разговорный) сразу пошла работать': 'B2',
    'Б1 ( проходит Б2 )': 'B1-B2', 'a0-a1': 'A0',
    'Б1 ( был екзамен ждет итог )': 'B1', 'Б2-С1': 'B2-C1',
    'b1 (учила, но не сдала В2)': 'B1', 'ня а2, ая в1': 'A2-B1',
    'A2 (идет на В1)': 'A2-B1', 'B2-C2': 'B2', 'немецкий - а1-а2, англ b1-b2': 'A2',
    'B2+': 'B2', 'в1, еще нет сертификата': 'B1', 'б1-б2': 'B1-B2', 'Бй': 'B1',
    'ждет результаты по B1 экзамену': 'B1', 'b2 (ждет серт)': 'B2', 'никакой': 'A0',
    'в1 , хочет совмещать с в2': 'B1-B2', 90: 'Неизвестно', '.': 'Неизвестно',
    'в1 (уже сдала В2)': 'B2', 'b1 результат экзамена в феврале': 'B1',
    'в1 , экзамен на в2 15 декабря': 'B1-B2', 'идет на А1': 'A1',
    'УТОЧНИТЬ!': 'Неизвестно', 'B2 (говорит без проблем - давно здесь)': 'B2',
    'B1 (до февраля)': 'B1',     'А2 ( Б2 в процессе)': 'B1', 'C': 'C1',
    'б1 заканчивает': 'B1', 'B1 (B2 экзамен в январе)': 'B1-B2',
    '5 июля 2024 сдает экз на В2': 'B2', 'А2 (заканчив В1 в июне)': 'A2-B1',
    'a2-б1': 'A2-B1', 'В1?': 'B1', 'b1 будет в январе экзамен, готов совмещать': 'B1',
    'b1 (b2 экзамен 2 марта)': 'B1-B2', 'B1 немецкий и английский Advance': 'B1',
    'A': 'A1', 'a2 (b1 экзамен 15 июня)': 'A2-B1', 'B2 (ждет итог экзамена)': 'B2',
    'b1 (b2 не сдал экзамен)': 'B1', 'В1 (учится на В2 до авг.': 'B1-B2',
    'В2 - не сдал': 'B2', 'B2+ (не сдавал, но говорит)': 'B2',
    'b1 (ждет серт)': 'B1', 'B1 вроде был (18 лет назад сдавал)': 'B1',
    'А2 (сдает B1 - 12 дек) - не сдал!': 'A2', 'УТОЧНИТЬ': 'Неизвестно',
    'b2 ждет серт': 'B2', 'разговорный из украины, без сертификата': 'B2',
    'Ждет B1': 'B1', 'сдавала А2 в сентябре': 'A2',
    'В1, учится на В2 до няоб 24': 'B1-B2',
    'Б1 ( ждет результат Б2)': 'B1-B2',
    'точно уровень не знаю, но говорить могу - учила сама': 'B2',
    'А2-В1 учит': 'A2-B1', 'В1 (учится на В2 уже)': 'B1',
    'В январе - В2 сдает': 'B2',
    'b1 должна получить результаты в феврале': 'B1'
}
deals_df['Level of Deutsch'] = deals_df['Level of Deutsch']\
.replace(replace_dict).fillna("Unknown")

In [ ]:
replace_take_two = {"B1-B2": "B2", 'A2-B1': 'B1', 'B2-C1': 'C1', 'A1-A2': "A2", "Неизвестно": "Unknown"}
deals_df['Level of Deutsch'] = deals_df['Level of Deutsch'].replace(replace_take_two)

In [ ]:
deals_df['Level of Deutsch'].unique()

array(['Unknown', 'B1', 'A2', 'B2', 'A0', 'C2', 'C1', 'A1'], dtype=object)

In [ ]:
deals_df['Course duration'].value_counts()

Course duration
11    3012
6      575
Name: count, dtype: Int64

In [ ]:
mode_duration = deals_df.groupby('Product')['Course duration'].agg(lambda x: x.mode()[0] if not x.mode().empty else None)
mode_duration

Product
Data Analytics         <NA>
Digital Marketing        11
Find yourself in IT    <NA>
UX/UI Design             11
Web Developer             6
Name: Course duration, dtype: Int8

#### ☑ The 'Digital Marketing' is the same course 'Data Analytics.' To ensure consistency and avoid confusion in analysis, I standardized the values by renaming all instances of 'Data Analytics' to 'Digital Marketing.

In [ ]:
deals_df['Product'] = deals_df['Product'].replace('Data Analytics', 'Digital Marketing')

In [ ]:
deals_df['Education Type'].value_counts()

Education Type
Morning    2895
Evening     404
Name: count, dtype: int64

In [ ]:
deals_df[deals_df['Education Type'].isna()][['Education Type', 'Course duration', 'Offer Total Amount']].isna().sum()


Education Type        18294
Course duration       17997
Offer Total Amount    17303
dtype: int64

In [ ]:
mode_education = deals_df.groupby('Product')['Education Type'].agg(lambda x: x.mode()[0] if not x.mode().empty else None)
mode_education

Product
Digital Marketing      Morning
Find yourself in IT       None
UX/UI Design           Morning
Web Developer          Morning
Name: Education Type, dtype: object

In [ ]:
mode_education = deals_df.groupby(
    ['Offer Total Amount', 'Course duration', 'Product']
)['Education Type'].agg(lambda x: x.mode().iloc[0] if not x.mode().empty else None)
mode_education

Offer Total Amount  Course duration  Product          
0.0                 11               Digital Marketing    Evening
1000.0              6                Web Developer        Morning
1200.0              11               Digital Marketing    Morning
1500.0              11               Digital Marketing    Morning
                                     UX/UI Design         Morning
2000.0              6                Web Developer        Morning
                    11               UX/UI Design         Evening
2500.0              6                Web Developer        Morning
                    11               Digital Marketing    Morning
                                     UX/UI Design         Evening
2900.0              6                Web Developer        Morning
                    11               Digital Marketing    Evening
                                     UX/UI Design         Morning
3000.0              11               Digital Marketing    Evening
                                     UX/UI Design         Morning
3500.0              6                Web Developer        Morning
                    11               Digital Marketing    Evening
                                     UX/UI Design         Evening
4000.0              11               Digital Marketing    Evening
                                     UX/UI Design         Evening
4500.0              11               Digital Marketing    Evening
                                     UX/UI Design         Evening
5000.0              6                Web Developer        Morning
                    11               Digital Marketing    Evening
6500.0              6                Web Developer        Morning
9000.0              6                Web Developer        Morning
                    11               UX/UI Design         Morning
10000.0             6                Web Developer        Morning
11000.0             6                Web Developer        Morning
                    11               Digital Marketing    Morning
                                     UX/UI Design         Morning
11111.0             11               Digital Marketing    Morning
11398.0             11               Digital Marketing    Morning
11500.0             11               Digital Marketing    Morning
                                     UX/UI Design         Morning
Name: Education Type, dtype: object

#### ☑ We fill missing values in 'Education Type' with the most frequent type for each combination of 'Offer Total Amount', 'Course duration', and 'Product'. This ensures uniformity in the data and improves accuracy for analysis.

In [ ]:
mode_education = deals_df.groupby(
    ['Offer Total Amount', 'Course duration', 'Product']
)['Education Type'].agg(lambda x: x.mode().iloc[0] if not x.mode().empty else None).reset_index()

deals_df = deals_df.merge(mode_education, on=['Offer Total Amount', 'Course duration', 'Product'], how='left', suffixes=('', '_mode'))

deals_df['Education Type'] = deals_df['Education Type'].fillna(deals_df['Education Type_mode'])

deals_df.drop(columns=['Education Type_mode'], inplace=True)

In [ ]:
deals_df['Education Type'].isna().sum()

18154

In [ ]:
deals_df.isna().sum()

Id                         0
Deal Owner Name           29
Closing Date            6948
Quality                 2253
Stage                      0
Lost Reason             5469
Page                       0
Campaign                5526
Content                 7446
Term                    9139
Source                     0
Payment Type           21097
Product                18001
Education Type         18154
Created Time               0
Course duration        18006
Months of study        20753
Initial Amount Paid    17428
Offer Total Amount     17408
Contact Name              61
City                   18307
Level of Deutsch           0
SLA Seconds             6060
dtype: int64

#### ☑ We fill missing values in 'Contact Name' fot 'Unknown'. That ensures that all data is retained during grouping and mapping, preventing errors and missing values. It also simplifies filtering and analysis by avoiding undefined entries.

In [ ]:
deals_df['Contact Name'] = deals_df['Contact Name'].fillna('Unknown')


In [ ]:
print(deals_df['Course duration'].unique())
print(deals_df['Months of study'].unique())

<IntegerArray>
[<NA>, 6, 11]
Length: 3, dtype: Int8
<IntegerArray>
[<NA>, 1, 0, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11]
Length: 13, dtype: Int8


In [ ]:
deals_df[deals_df['Deal Owner Name'].isna()][['Deal Owner Name', 'Offer Total Amount', 'Initial Amount Paid']].isna().sum()

Deal Owner Name        29
Offer Total Amount     13
Initial Amount Paid    13
dtype: int64

In [ ]:
deals_df[deals_df['Deal Owner Name'].isna()][['Deal Owner Name', 'Offer Total Amount', 'Initial Amount Paid']]

Deal Owner Name  Offer Total Amount  Initial Amount Paid
14007             NaN                 0.0                  0.0
18834             NaN                 NaN                  NaN
18954             NaN                 NaN                  NaN
19054             NaN                 NaN                  NaN
19150             NaN                 NaN                  NaN
19248             NaN                 NaN                  NaN
19254             NaN                 NaN                  NaN
19337             NaN                 NaN                  NaN
19382             NaN                 NaN                  NaN
19438             NaN                 NaN                  NaN
19439             NaN                 NaN                  NaN
19482             NaN                 NaN                  NaN
19727             NaN                 NaN                  NaN
19728             NaN                 NaN                  NaN
21357             NaN                 0.0                  0.0
21385             NaN                 0.0                  0.0
21416             NaN                 0.0                  0.0
21424             NaN                 0.0                  0.0
21432             NaN                 0.0                  0.0
21440             NaN                 0.0                  0.0
21451             NaN                 0.0                  0.0
21463             NaN                 0.0                  0.0
21518             NaN                 0.0                  0.0
21527             NaN                 0.0                  0.0
21531             NaN                 0.0                  0.0
21538             NaN                 0.0                  0.0
21544             NaN                 0.0                  0.0
21560             NaN                 0.0                  0.0
21568             NaN                 0.0                  0.0

#### ☑ We remove rows where Deal Owner Name is Nan, becouse, Offer Total Amount and Initial Amount Paid are all zero in them, as they do not provide useful information and may distort data analysis.

In [ ]:
deals_df = deals_df.dropna(subset=['Deal Owner Name'])

In [ ]:
deals_df['Deal Owner Name'].isna().sum()

0

In [ ]:
deals_df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 21564 entries, 0 to 21592
Data columns (total 23 columns):
 #   Column               Non-Null Count  Dtype         
---  ------               --------------  -----         
 0   Id                   21564 non-null  object        
 1   Deal Owner Name      21564 non-null  object        
 2   Closing Date         14638 non-null  datetime64[ns]
 3   Quality              19311 non-null  object        
 4   Stage                21564 non-null  object        
 5   Lost Reason          16118 non-null  object        
 6   Page                 21564 non-null  object        
 7   Campaign             16040 non-null  object        
 8   Content              14123 non-null  object        
 9   Term                 12431 non-null  object        
 10  Source               21564 non-null  object        
 11  Payment Type         495 non-null    object        
 12  Product              3592 non-null   object        
 13  Education Type       3439 non-null  

In [ ]:
deals_df['Quality'].value_counts()

Quality
E - Non Qualified    7633
D - Non Target       6248
C - Low              3442
B - Medium           1556
A - High              429
F                       3
Name: count, dtype: int64

In [ ]:
deals_df = deals_df[deals_df['Quality'] != 'F']

---
### Removing duplicates
---

In [ ]:
deals_df = deals_df[deals_df['Lost Reason'] != 'Duplicate']

In [ ]:
deals_df.duplicated(subset=deals_df.columns[1:]).sum()

5

In [ ]:
deals_df[deals_df.duplicated(keep=False, subset=deals_df.columns[1:])]

Id Deal Owner Name Closing Date Quality  \
1067   5805028000054269273        John Doe          NaT     NaN   
1069   5805028000054269205        John Doe          NaT     NaN   
11084  5805028000029580838   Charlie Davis          NaT     NaN   
11085  5805028000029576850   Charlie Davis          NaT     NaN   
14754  5805028000020283932   Charlie Davis          NaT     NaN   
14755  5805028000020420273   Charlie Davis          NaT     NaN   
14807  5805028000020313482     Diana Evans          NaT     NaN   
14808  5805028000020284716     Diana Evans          NaT     NaN   
15534  5805028000018428977      Nina Scott          NaT     NaN   
15535  5805028000018507379      Nina Scott          NaT     NaN   

                       Stage Lost Reason       Page                 Campaign  \
1067                New Lead         NaN   /webinar                      NaN   
1069                New Lead         NaN   /webinar                      NaN   
11084  Registered on Webinar         NaN  /workshop                      NaN   
11085  Registered on Webinar         NaN  /workshop                      NaN   
14754  Registered on Webinar         NaN  /workshop                      NaN   
14755  Registered on Webinar         NaN  /workshop                      NaN   
14807  Registered on Webinar         NaN  /workshop                      NaN   
14808  Registered on Webinar         NaN  /workshop                      NaN   
15534  Registered on Webinar         NaN  /workshop  14.11.23wide_webinar_DE   
15535  Registered on Webinar         NaN  /workshop  14.11.23wide_webinar_DE   

                     Content  Term  ... Education Type        Created Time  \
1067                     NaN   NaN  ...            NaN 2024-06-10 12:37:00   
1069                     NaN   NaN  ...            NaN 2024-06-10 12:37:00   
11084                    NaN   NaN  ...            NaN 2024-02-01 20:45:00   
11085                    NaN   NaN  ...            NaN 2024-02-01 20:45:00   
14754             invitation   NaN  ...            NaN 2023-12-08 22:06:00   
14755             invitation   NaN  ...            NaN 2023-12-08 22:06:00   
14807             invitation   NaN  ...            NaN 2023-12-08 15:44:00   
14808             invitation   NaN  ...            NaN 2023-12-08 15:44:00   
15534  bloggersvideo8webinar  wide  ...            NaN 2023-11-23 14:06:00   
15535  bloggersvideo8webinar  wide  ...            NaN 2023-11-23 14:06:00   

      Course duration Months of study Initial Amount Paid  Offer Total Amount  \
1067             <NA>            <NA>                 NaN                 NaN   
1069             <NA>            <NA>                 NaN                 NaN   
11084            <NA>            <NA>                 NaN                 NaN   
11085            <NA>            <NA>                 NaN                 NaN   
14754            <NA>            <NA>                 NaN                 NaN   
14755            <NA>            <NA>                 NaN                 NaN   
14807            <NA>            <NA>                 NaN                 NaN   
14808            <NA>            <NA>                 NaN                 NaN   
15534            <NA>            <NA>                 NaN                 NaN   
15535            <NA>            <NA>                 NaN                 NaN   

              Contact Name  City  Level of Deutsch SLA Seconds  
1067   5805028000005448163  None           Unknown         NaN  
1069   5805028000005448163  None           Unknown         NaN  
11084  5805028000025816321  None           Unknown         NaN  
11085  5805028000025816321  None           Unknown         NaN  
14754  5805028000020421099  None           Unknown         NaN  
14755  5805028000020421099  None           Unknown         NaN  
14807  5805028000020300454  None           Unknown         NaN  
14808  5805028000020300454  None           Unknown         NaN  
15534  5805028000018427915  None           Unknown         NaN  
1553

In [ ]:
deals_df.drop_duplicates(subset=deals_df.columns[1:], inplace=True)

deals_df[deals_df.duplicated(subset=deals_df.columns[1:])].shape

(0, 23)

In [ ]:
deals_df[deals_df['Offer Total Amount'].isna()][['Offer Total Amount', 'Initial Amount Paid', 'Months of study']]

Offer Total Amount  Initial Amount Paid  Months of study
0                     NaN                  NaN             <NA>
2                     NaN                  NaN             <NA>
3                     NaN                  NaN             <NA>
4                     NaN                  NaN             <NA>
5                     NaN                  NaN             <NA>
...                   ...                  ...              ...
21584                 NaN                  NaN             <NA>
21587                 NaN                  NaN             <NA>
21589                 NaN                  NaN             <NA>
21590                 NaN                  NaN             <NA>
21591                 NaN                  NaN             <NA>

[15724 rows x 3 columns]

In [ ]:
category_columns = ['Deal Owner Name', 'Quality', 'Stage', 'Lost Reason', 'Page', 'Campaign', 'Content',
                     'Term', 'Source', 'Payment Type', 'Product', 'City', 'Level of Deutsch', 'Education Type']
for col in category_columns:
  deals_df[col] = deals_df[col].astype('category')

---
### Results Output
---

In [ ]:
deals_df.head()

Id Deal Owner Name Closing Date            Quality  \
0  5805028000056864695        Ben Hall          NaT                NaN   
1  5805028000056859489   Ulysses Adams          NaT                NaN   
2  5805028000056832357   Ulysses Adams   2024-06-21     D - Non Target   
3  5805028000056824246        Eva Kent   2024-06-21  E - Non Qualified   
4  5805028000056873292        Ben Hall   2024-06-21     D - Non Target   

      Stage     Lost Reason       Page                  Campaign  \
0  New Lead             NaN  /eng/test             03.07.23women   
1  New Lead             NaN    /at-eng                       NaN   
2      Lost      Non target    /at-eng                engwien_AT   
3      Lost  Invalid number       /eng  04.07.23recentlymoved_DE   
4      Lost      Non target       /eng              discovery_DE   

              Content           Term  ... Education Type        Created Time  \
0                 v16          women  ...            NaN 2024-06-21 15:30:00   
1                 NaN            NaN  ...        Morning 2024-06-21 15:23:00   
2               b1-at     21_06_2024  ...            NaN 2024-06-21 14:45:00   
3  bloggersvideo14com  recentlymoved  ...            NaN 2024-06-21 13:32:00   
4             website            NaN  ...            NaN 2024-06-21 13:21:00   

  Course duration Months of study Initial Amount Paid  Offer Total Amount  \
0            <NA>            <NA>                 NaN                 NaN   
1               6            <NA>                 0.0              2000.0   
2            <NA>            <NA>                 NaN                 NaN   
3            <NA>            <NA>                 NaN                 NaN   
4            <NA>            <NA>                 NaN                 NaN   

          Contact Name  City  Level of Deutsch SLA Seconds  
0  5805028000056849495   NaN           Unknown         NaN  
1  5805028000056834471   NaN           Unknown         NaN  
2  5805028000056854421   NaN           Unknown      1603.0  
3  5805028000056889351   NaN           Unknown      3604.0  
4  5805028000056876176   NaN           Unknown      3192.0  

[5 rows x 23 columns]

In [ ]:
deals_df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 19788 entries, 0 to 21592
Data columns (total 23 columns):
 #   Column               Non-Null Count  Dtype         
---  ------               --------------  -----         
 0   Id                   19788 non-null  object        
 1   Deal Owner Name      19788 non-null  category      
 2   Closing Date         13129 non-null  datetime64[ns]
 3   Quality              17557 non-null  category      
 4   Stage                19788 non-null  category      
 5   Lost Reason          14347 non-null  category      
 6   Page                 19788 non-null  category      
 7   Campaign             15555 non-null  category      
 8   Content              13771 non-null  category      
 9   Term                 12077 non-null  category      
 10  Source               19788 non-null  category      
 11  Payment Type         482 non-null    category      
 12  Product              3537 non-null   category      
 13  Education Type       3390 non-null  

In [ ]:
deals_df.describe()

Closing Date                   Created Time  \
count                          13129                          19788   
mean   2024-01-28 16:28:52.957574656  2024-01-28 16:11:33.999393536   
min              2022-10-11 00:00:00            2023-07-03 17:03:00   
25%              2023-11-11 00:00:00            2023-11-18 23:15:30   
50%              2024-02-08 00:00:00            2024-02-06 16:15:00   
75%              2024-04-19 00:00:00            2024-04-14 23:28:45   
max              2024-12-11 00:00:00            2024-06-21 15:30:00   
std                              NaN                            NaN   

       Course duration  Months of study  Initial Amount Paid  \
count           3533.0            838.0          4044.000000   
mean         10.197566         5.445107           961.006429   
min                6.0              0.0             0.000000   
25%               11.0              3.0           300.000000   
50%               11.0              5.0          1000.000000   
75%               11.0              8.0          1000.000000   
max               11.0             11.0         11500.000000   
std           1.835545         2.918031          1417.220439   

       Offer Total Amount   SLA Seconds  
count         4064.000000  1.498000e+04  
mean          7238.117618  1.115745e+05  
min              0.000000  3.000000e+00  
25%           3500.000000  4.408250e+03  
50%          11000.000000  1.982000e+04  
75%          11000.000000  5.596900e+04  
max          11500.000000  2.690846e+07  
std           4584.678975  7.268762e+05

In [ ]:
deals_df

Id Deal Owner Name Closing Date            Quality  \
0      5805028000056864695        Ben Hall          NaT                NaN   
1      5805028000056859489   Ulysses Adams          NaT                NaN   
2      5805028000056832357   Ulysses Adams   2024-06-21     D - Non Target   
3      5805028000056824246        Eva Kent   2024-06-21  E - Non Qualified   
4      5805028000056873292        Ben Hall   2024-06-21     D - Non Target   
...                    ...             ...          ...                ...   
21587  5805028000000935025    Kevin Parker   2023-07-06  E - Non Qualified   
21589  5805028000000948010      Jane Smith   2023-08-29         B - Medium   
21590  5805028000000945016      Jane Smith   2023-08-29           A - High   
21591  5805028000000927004       Bob Brown   2023-07-09     D - Non Target   
21592  5805028000000922001       Bob Brown   2023-07-03  E - Non Qualified   

          Stage             Lost Reason                   Page  \
0      New Lead                     NaN              /eng/test   
1      New Lead                     NaN                /at-eng   
2          Lost              Non target                /at-eng   
3          Lost          Invalid number                   /eng   
4          Lost              Non target                   /eng   
...         ...                     ...                    ...   
21587      Lost          Doesn't Answer  eng/digital-marketing   
21589      Lost     needs time to think  eng/digital-marketing   
21590      Lost        Changed Decision  eng/digital-marketing   
21591      Lost  Does not speak English  eng/digital-marketing   
21592      Lost                 Refugee                      /   

                       Campaign             Content           Term  ...  \
0                 03.07.23women                 v16          women  ...   
1                           NaN                 NaN            NaN  ...   
2                    engwien_AT               b1-at     21_06_2024  ...   
3      04.07.23recentlymoved_DE  bloggersvideo14com  recentlymoved  ...   
4                  discovery_DE             website            NaN  ...   
...                         ...                 ...            ...  ...   
21587           02.07.23wide_DE                  b3           wide  ...   
21589             03.07.23women                  b3          women  ...   
21590           02.07.23wide_DE                  b3           wide  ...   
21591             03.07.23women                  b3          women  ...   
21592                       NaN                 NaN            NaN  ...   

      Education Type        Created Time Course duration Months of study  \
0                NaN 2024-06-21 15:30:00            <NA>            <NA>   
1            Morning 2024-06-21 15:23:00               6            <NA>   
2                NaN 2024-06-21 14:45:00            <NA>            <NA>   
3                NaN 2024-06-21 13:32:00            <NA>            <NA>   
4                NaN 2024-06-21 13:21:00            <NA>            <NA>   
...              ...                 ...             ...             ...   
21587            NaN 2023-07-04 09:34:00            <NA>            <NA>   
21589            NaN 2023-07-04 07:10:00            <NA>            <NA>   
21590            NaN 2023-07-03 20:39:00            <NA>            <NA>   
21591            NaN 2023-07-03 20:17:00            <NA>            <NA>   
21592            NaN 2023-07-03 17:03:00            <NA>            <NA>   

      Initial Amount Paid  Offer Total Amount         Contact Name  City  \
0                     NaN                 NaN  5805028000056849495   NaN   
1                     0.0              2000.0  5805028000056834471   NaN   
2                     NaN                 NaN  5805028000056854421   NaN   
3                     NaN                 NaN  5805028000056889351   NaN   
4                     NaN                 NaN  5805028000056876176   NaN   
...                   ..

In [ ]:
mask = deals_df['Initial Amount Paid'] > deals_df['Offer Total Amount']
if mask.any():
    deals_df.loc[mask, ['Initial Amount Paid', 'Offer Total Amount']] = \
        deals_df.loc[mask, ['Offer Total Amount', 'Initial Amount Paid']].values

print(deals_df[deals_df['Offer Total Amount'] < deals_df['Initial Amount Paid']])


Empty DataFrame
Columns: [Id, Deal Owner Name, Closing Date, Quality, Stage, Lost Reason, Page, Campaign, Content, Term, Source, Payment Type, Product, Education Type, Created Time, Course duration, Months of study, Initial Amount Paid, Offer Total Amount, Contact Name, City, Level of Deutsch, SLA Seconds]
Index: []

[0 rows x 23 columns]


In [ ]:
deals_df.to_excel("Deals (Result).xlsx", index=False)



---



# **Descriptive Statistics:**
1. Calculate summary statistics (mean, median, mode, range) for numeric fields.
2. Analyze categorical fields such as quality, stage, source, and product.



---



In [ ]:
contacts_df.describe()

Created Time                  Modified Time
count                          18510                          18510
mean   2024-01-24 14:25:56.155591680  2024-02-15 09:11:05.273905920
min              2023-06-27 11:28:00            2023-07-06 10:54:00
25%              2023-11-15 16:49:15            2023-12-09 14:51:45
50%              2024-02-01 18:44:30            2024-02-29 01:12:30
75%              2024-04-12 16:15:45            2024-04-26 22:40:45
max              2024-06-21 15:30:00            2024-06-21 15:32:00

### **Descriptive Statistics Insights:**

1. **Data Volume**: 18,510 records, indicating a significant number of events included in the analysis.

2. **Mean Values**:  
   - Average creation date: January 24, 2024.  
   - Average modification date: February 15, 2024.  
   This suggests that, on average, records are modified approximately 21 days after creation.

3. **Min and Max Values**:  
   - The earliest record was created on June 27, 2023, and the latest on June 21, 2024.  
   - The first modification occurred on July 6, 2023, and the last on June 21, 2024.  
   The data covers a span of almost one year.

4. **Median Values (50th Percentile)**:  
   - Half of the records were created before February 1, 2024.  
   - Half of the modifications occurred before February 29, 2024.  
   This indicates high activity in record creation during late 2023 and early 2024.

5. **25th and 75th Percentiles**:  
   - 25% of records were created before November 15, 2023, and 75% before April 12, 2024.  
   - For modifications, the corresponding dates are December 9, 2023, and April 26, 2024.  
   This suggests a relatively even distribution with a slight skew toward the end of the period.

6. **Time Between Creation and Modification**:  
   - On average, records are modified a few weeks after creation.  
   - The spread of values indicates that some records are updated quickly, while others remain unchanged for extended periods.

### **Conclusion: **
The data shows steady growth in activity, peaking in early 2024. The average time to modify a record is a few weeks, which may indicate either frequent need for data adjustments or a workflow that involves refining records after their initial creation.

In [ ]:
deals_df.describe()

Closing Date                   Created Time  \
count                          13129                          19788   
mean   2024-01-28 16:28:52.957574656  2024-01-28 16:11:33.999393536   
min              2022-10-11 00:00:00            2023-07-03 17:03:00   
25%              2023-11-11 00:00:00            2023-11-18 23:15:30   
50%              2024-02-08 00:00:00            2024-02-06 16:15:00   
75%              2024-04-19 00:00:00            2024-04-14 23:28:45   
max              2024-12-11 00:00:00            2024-06-21 15:30:00   
std                              NaN                            NaN   

       Course duration  Months of study  Initial Amount Paid  \
count           3533.0            838.0          4044.000000   
mean         10.197566         5.445107           940.479970   
min                6.0              0.0             0.000000   
25%               11.0              3.0           300.000000   
50%               11.0              5.0          1000.000000   
75%               11.0              8.0          1000.000000   
max               11.0             11.0         11000.000000   
std           1.835545         2.918031          1341.419686   

       Offer Total Amount   SLA Seconds  
count         4064.000000  1.498000e+04  
mean          7258.543061  1.115745e+05  
min              0.000000  3.000000e+00  
25%           3500.000000  4.408250e+03  
50%          11000.000000  1.982000e+04  
75%          11000.000000  5.596900e+04  
max          11500.000000  2.690846e+07  
std           4579.304930  7.268762e+05

### Key Insights from the Data:  

1. **Dates**:  
   - Average deal closing date: January 28, 2024.  
   - Earliest closing: October 11, 2022; latest: December 11, 2024.  
   - Average deal creation date: January 28, 2024 (almost matches the average closing date).  

2. **Course Duration**:  
   - Average course duration: 10.2 months; median: 11 months.  
   - Minimum duration: 6 months; maximum: 11 months.  
   - Distribution is tight, with most courses lasting 11 months.  

3. **Financial Metrics**:  
   - Average initial payment: 961 USD; median: 1,000 USD.  
   - Average offer amount: 7,238 USD; median: 11,000 USD.  
   - Minimum payment: 0 USD; maximum: 11,500 USD, indicating possible trial/free courses.  

4. **SLA (Deal Processing Time in Seconds)**:  
   - Average processing time: 111,574 seconds (~31 hours).  
   - Median: 19,820 seconds (~5.5 hours), indicating strong asymmetry.  
   - Maximum value: 26,908,460 seconds (~311 days) — potential outliers.  

### Overall Conclusion:  
- Most courses last 11 months, with a high average cost of 7,238 USD.  
- Some deals were processed for free or with minimal upfront payment.  
- Processing time varies significantly, with some extreme outliers.  
- SLA analysis requires further investigation into anomalously high values.

In [ ]:
spend_df.describe()

Date    Impressions         Spend  \
count                          14967   14967.000000  14967.000000   
mean   2024-01-09 18:42:18.544798464    3411.951092      9.970728   
min              2023-07-03 00:00:00       0.000000      0.000000   
25%              2023-10-07 00:00:00      35.000000      0.310000   
50%              2024-01-20 00:00:00     286.000000      2.350000   
75%              2024-04-11 00:00:00    1057.000000      8.890000   
max              2024-06-21 00:00:00  431445.000000    774.000000   
std                              NaN   13361.336847     31.077729   

             Clicks  
count  14967.000000  
mean      30.978686  
min        0.000000  
25%        0.000000  
50%        4.000000  
75%       17.000000  
max     2415.000000  
std       98.175199

### **Insights from Descriptive Statistics of Advertising Data:**

1. **Data Volume**:  
   - The sample contains 14,967 records, representing a significant amount of information on ad impressions, costs, and clicks.

2. **Mean Values**:  
   - Average ad display date: January 9, 2024.  
   - Per record, there are on average 3,411 impressions, 9.97 monetary units spent, and 30.98 clicks.  
   - However, high standard deviations (especially for impressions: 13,361; clicks: 98.18) indicate significant data variability.

3. **Min and Max Values**:  
   - Earliest ad display: July 3, 2023; latest: June 21, 2024.  
   - Minimum values for all metrics are 0 (some records have no impressions, clicks, or costs).  
   - Maximum values:  
     - Impressions: 431,445 (a huge gap from the median, indicating rare outliers).  
     - Costs: 774 units.  
     - Clicks: 2,415.  

4. **Percentile Analysis (Value Distribution)**:  
   - 25% of records have 35 or fewer impressions, costs up to 0.31, and 0 clicks (a significant portion of campaigns receive no engagement).  
   - 50% (median) records have 286 impressions, 2.35 costs, and 4 clicks – a more realistic benchmark for typical campaigns.  
   - 75% of records receive up to 1,057 impressions, 8.89 costs, and 17 clicks – above average but far from maximum values.  

5. **Data Spread and Features**:  
   - The distribution is uneven, with many records showing low metrics but rare records with extremely high impressions, costs, and clicks.  
   - A large share of records (25%) receive no clicks, indicating low audience engagement for some ads.  
   - Mean values (3,411 impressions, 9.97 costs, 30.98 clicks) are inflated by rare spikes in activity.

### *Conclusion:*  
- Most records show low impressions, costs, and clicks, but rare records with anomalously high values skew the averages.  
- Median values (286 impressions, 2.35 costs, 4 clicks) better reflect the real picture than the means.  
- A quarter of records receive no clicks, suggesting ineffective ads or poor targeting.  
- High data variance indicates that some campaigns significantly outperform others in reach and engagement.

In [ ]:
calls_df.describe()

Call Start Time  Call Duration (in seconds)
count                          92617                92538.000000
mean   2024-02-04 19:21:56.171329024                  170.330178
min              2023-06-30 08:43:00                    0.000000
25%              2023-11-23 17:14:00                    4.000000
50%              2024-02-16 16:00:00                    9.000000
75%              2024-04-22 14:39:00                  107.000000
max              2024-06-21 15:31:00                 7625.000000
std                              NaN                  406.944468

### **Insights from Descriptive Statistics of Call Data:**

1. **Data Volume**:  
   - Total recorded calls: 92,617.  
   - Call duration is available for 92,538 records*, indicating a small number of calls with missing duration data.

2. **Mean Values**:  
   - Average call start time: February 4, 2024.  
   - Average call duration: 170 seconds (2 minutes 50 seconds).  
   - However, the high standard deviation (406.94 seconds) indicates significant variability in call lengths.

3. **Min and Max Values**:  
   - Earliest call in the dataset: June 30, 2023; latest: June 21, 2024, covering almost 1 year.  
   - Minimum call duration: 0 seconds (possibly unanswered calls).  
   - Maximum call duration: 7,625 seconds (~2 hours 7 minutes), indicating rare but very long calls.

4. **Percentile Analysis (Duration Distribution)**:  
   - 25% of calls lasted up to 4 seconds — very short interactions (possibly missed calls or hang-ups).  
   - 50% (median) of calls lasted 9 seconds, highlighting a high proportion of short calls.  
   - 75% of calls lasted up to 107 seconds (1 minute 47 seconds), bringing the average closer to more typical call durations.  

5. **Data Spread and Features**:  
   - Most calls are short (under 10 seconds), but there is a significant number of longer calls, which inflates the average duration.  
   - High standard deviation suggests the presence of rare, extremely long calls that heavily influence the mean.

### *Conclusion: *
- Most calls are short, with a median duration of 9 seconds.  
- Around 25% of calls last less than 4 seconds, potentially indicating frequent missed or dropped calls.  
- Long calls (over 7,000 seconds) are rare but significantly increase the average duration.  
- The data shows a trend of many short calls with a small number of very long ones.

## 🏷️ **Category**

#-- Counting unique values --

In [ ]:
categorical_columns = ['Quality', 'Stage', 'Source', 'Product']

for col in categorical_columns:
    print(f"Уникальные значения в {col}:")
    print(deals_df[col].unique(), "\n")

Уникальные значения в Quality:
[NaN, 'D - Non Target', 'E - Non Qualified', 'B - Medium', 'C - Low', 'A - High']
Categories (5, object): ['A - High', 'B - Medium', 'C - Low', 'D - Non Target',
                         'E - Non Qualified'] 

Уникальные значения в Stage:
['New Lead', 'Lost', 'Need a consultation', 'Need To Call', 'Call Delayed', ..., 'Need to Call - Sales', 'Test Sent', 'Payment Done', 'Registered on Offline Day', 'Free Education']
Length: 13
Categories (13, object): ['Call Delayed', 'Free Education', 'Lost', 'Need To Call', ...,
                          'Registered on Offline Day', 'Registered on Webinar', 'Test Sent',
                          'Waiting For Payment'] 

Уникальные значения в Source:
['Facebook Ads', 'Organic', 'Telegram posts', 'Google Ads', 'Youtube Ads', ..., 'Bloggers', 'Webinar', 'Partnership', 'Test', 'Offline']
Length: 13
Categories (13, object): ['Bloggers', 'CRM', 'Facebook Ads', 'Google Ads', ..., 'Test', 'Tiktok Ads',
                         

# -- Counting the frequencies of occurrence --

In [ ]:
for col in categorical_columns:
    print(f"Распределение значений в {col}:")
    print(deals_df[col].value_counts(), "\n")

Распределение значений в Quality:
Quality
E - Non Qualified    6143
D - Non Target       6078
C - Low              3386
B - Medium           1535
A - High              415
Name: count, dtype: int64 

Распределение значений в Stage:
Stage
Lost                         13991
Call Delayed                  2243
Registered on Webinar         2066
Payment Done                   855
Waiting For Payment            324
Qualificated                   106
Registered on Offline Day       85
Need to Call - Sales            33
Need To Call                    31
Test Sent                       25
Need a consultation             23
New Lead                         5
Free Education                   1
Name: count, dtype: int64 

Распределение значений в Source:
Source
Facebook Ads      4711
Google Ads        4110
Tiktok Ads        2001
SMM               1668
Youtube Ads       1618
Organic           1497
CRM               1455
Bloggers          1071
Telegram posts     993
Webinar            305
Partnersh

# -- Percentage Distribution --

In [ ]:
deals_df['Quality'].value_counts(normalize=True) * 100

Quality
E - Non Qualified    34.988893
D - Non Target       34.618671
C - Low              19.285755
B - Medium            8.742952
A - High              2.363730
Name: proportion, dtype: float64

# -- Analysis of the Relationship Between Categories and Numerical Variables --

In [ ]:
for col in categorical_columns:
    print(deals_df.groupby(col, observed=True)['Offer Total Amount'].mean(), "\n")

Quality
A - High             7899.444444
B - Medium           8027.845884
C - Low              7790.862004
D - Non Target       8047.854027
E - Non Qualified     847.480106
Name: Offer Total Amount, dtype: float64 

Stage
Call Delayed                  9195.379781
Free Education                4000.000000
Lost                          6523.138811
Need To Call                          NaN
Need a consultation                   NaN
Need to Call - Sales                  NaN
New Lead                      2000.000000
Payment Done                  7547.739286
Qualificated                  5733.333333
Registered on Offline Day     5000.000000
Registered on Webinar         4000.000000
Test Sent                    10666.666667
Waiting For Payment          10084.829721
Name: Offer Total Amount, dtype: float64 

Source
Bloggers           8679.005525
CRM                8864.634146
Facebook Ads       7281.619820
Google Ads         6225.492341
Offline           11000.000000
Organic            8203.396